---

title: "RegisteredEventListeners.unregisterListener(CacheEventListener cacheEventListener) throws  java.lang.UnsupportedOperationException"
layout: issue
tags: 
permalink: /browse/EHC-653

issue_key: EHC-653
issue_numeric_sort_key: 653
issuetype: "Bug"
project: "Ehcache Core"
project_key: "EHC"
status: "Closed"
resolution: "Fixed"
priority: "1 Critical"
components: "ehcache-core"
labels: 
assignee: "hsingh"
reporter: "jbarbero"
votes:  0
watchers: 1

created: "2010-03-16T19:19:17.000-0400"
updated: "2013-01-17T18:32:32.000-0500"
resolved: "2010-03-17T06:43:26.000-0400"




---

{% raw %}

## Description

<div markdown="1" class="description">

Credit to Luciano Gregorio ( luciano.gregorio (write at) renxo com )

Recently there was a change in  net.sf.ehcache.event.RegisteredEventListeners in the 2.0 ehcache version.

The field RegisteredEventListeners.cacheEventListeners  now use CopyOnWriteArraySet like a set implementation.


public class RegisteredEventListeners \{

private final Set<ListenerWrapper> cacheEventListeners = new CopyOnWriteArraySet<ListenerWrapper>();
...
\}

The problem is that CopyOnWriteArraySet use an instance of CopyOnWriteArrayList  to iterate over the listeners, and CopyOnWriteArrayList does not support the remove method, so it throws an exception java.lang.UnsupportedOperationException .

This happen when it tries to unregister an event listener.

public final boolean unregisterListener(CacheEventListener cacheEventListener) \{
       Iterator<ListenerWrapper> it = cacheEventListeners.iterator();
       while (it.hasNext()) {
             ListenerWrapper listenerWrapper = it.next();
             if (listenerWrapper.getListener().equals(cacheEventListener)) {
                            // This call throws the UnsupportedOperationException //
                           it.remove();
                           return true;
            }
        }
        return false;
\}

This happend when you call CacheManager.shutdown() for example.

Thread => http://forums.terracotta.org/forums/posts/list/3304.page




</div>

## Comments


{:.comment-heading}
### **Geert Bevin** <span class="date">2010-03-17</span>

<div markdown="1" class="comment">

FWIW, the reason why this was not caught before was because CacheEventListenerTest just logged the exception without failing the tests. I added a new test for just this now.

</div>


{:.comment-heading}
### **Himadri Singh** <span class="date">2010-04-07</span>

<div markdown="1" class="comment">

CacheEventListenerTest is running on monkeys and not failing.

</div>



{% endraw %}
